### DISCUSSION TITLE
When max-pods count is set on a node in k3s cluster, how to get the list of pods (or) pod status that are considered in that limit?

### DISCUSSION THREAD

**Discussion by Koulini** [k3s-io/k3s#8286] [Q&A]
I have a system with max-pods count set to 100 on a node.There are pods with different status like Running/NodeAffinity/Completed/Terminating etc on the system. 

Observed that pods with NodeAffinity and Completed status are not considered in the pods count. 

1.Is there a way to get the list of pods that are considered by k3s for limiting the pods? 
2.Does k3s limit the creation of new pods by counting the pods with certain status like Running/Unknown? If yes, what statuses will be considered?

$ k3s --version
k3s version v1.21.1+k3s-2d1556d1-dirty (2d1556d1)
go version go1.15.2

**Comment by brandond** [comment]
This is really a Kubernetes question, not K3s. K3s does not have any custom logic for counting pods; this code is part of core Kubernetes. As far as I know, all of the pods that have been admitted to a node, count against that node's limit.

**Reply by Koulini** [reply]
I configured max-pods of node to 100.But i see more than 100 pods. 31 pods are in NodeAffinity state. 50 pods are in UnexpectedAdmissionError state. Pods in Running+Terminating state count to 100. Other pods stuck in Pending state as the pod limit is exceeding.

Events of pod in Pending state:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  19s   default-scheduler  0/1 nodes are available: 1 Too many pods.
  Warning  FailedScheduling  18s   default-scheduler  0/1 nodes are available: 1 Too many pods.
  
  Was able to create Pod when Running pod count was less than 100. Once pods in running state crossed 100, then pods are stuck in pending state.

**Reply by brandond** [reply]
Max pods doesn't limit how many you can create, it just limits how many can be scheduled to the node. Were you expecting the apiserver to reject your node creation once you hit that limit?

**Reply by brandond** [reply]
https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/
> --max-pods int32     Default: 110
>   Number of Pods that can run on this kubelet.

Note that it doesn't say anything about the number of pods that can be created, just the number that can run on the kubelet.

**Reply by Koulini** [reply]
I accept the behaviour of moving pods to Pending state,  as the pods limit is exceeding the configured value.I agree that there is no restriction for creation of pods.

Whats the expected behaviour of k3s when system is in below state?
If the system have 30 pods in NodeAffinity state and 70 pods in Running state and max-pods count set to 100 on node. Now a new pod is created. Will kubelet move the new pod to Running state as there are only 70 running pods on the system (or) to Pending state as total 100 pods are already exists on the system?

My usecase is to add  a check in my service to restrict the new pods creation, when the pods count on system matches the max limit allowed on node.

**Reply by brandond** [reply]
You should probably read the Kubernetes docs.
https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/

**Reply by Koulini** [reply]
All pods are of same priority..I want to know the k3s behaviour in above scenario i mentioned.

**Reply by brandond** [reply]
K3s is a Kubernetes distribution. It does whatever any other distribution of Kubernetes will do, given the same default scheduler configuration. The Kubernetes docs are the best source of info on scheduling behavior.
